File: src/parsing/operations/merge.ts
import Bindings from "../../graph/bindings";
import ASTNode from "../ast_node";
import Expression from "../expressions/expression";
import Operation from "./operation";

/**
 * One item in the `WHEN MATCHED THEN UPDATE SET` list.
 *
 * `field` is the name of the field on the matched target row to
 * overwrite; `expression` is the value to assign.  When `expression`
 * is `null`, the default is `source.<field>` — i.e. the value of the
 * same-named field on the current incoming row.  This makes the
 * partial-merge case (`SET .name, .email`) ergonomic without having to
 * write `SET .name = source.name, .email = source.email`.
 */
export interface MergeSetItem {
    field: string;
    expression: Expression | null;
}

/**
 * The action taken when an incoming source row matches an existing
 * target row.  Either rewrite specific fields (`update`) or remove
 * the matched target row (`delete`).
 */
export type MergeMatchedAction = { type: "update"; setItems: MergeSetItem[] } | { type: "delete" };

/**
 * The action taken when an incoming source row matches no existing
 * target row.  Inserts the row produced by `expression`, or the bare
 * source row if `expression` is `null` (the implicit-insert default).
 */
export interface MergeNotMatchedAction {
    type: "insert";
    expression: Expression | null;
}

/**
 * The `ON` clause: either a short-form list of key fields (matched
 * positionally against the source row by name) or an arbitrary
 * predicate evaluated per (target, source) pair.
 */
export type MergeOnClause =
    | { type: "keys"; keys: string[] }
    | { type: "predicate"; predicate: Expression };

/**
 * `MERGE INTO <name> [AS <target>] USING <source-rhs> [AS <source>]
 *      ON ( <key> | (<k1>, <k2>, …) | <predicate> )
 *      [ WHEN MATCHED THEN UPDATE SET <set-list> ]
 *      [ WHEN MATCHED THEN DELETE ]
 *      [ WHEN NOT MATCHED THEN INSERT [<expr>] ]`
 *
 * SQL-MERGE-style upsert into an existing list-binding.  The source
 * may be any expression, sub-query, or bare binding name; both the
 * `target` and `source` aliases are in scope inside the `ON`
 * predicate and any `SET .field = <expr>` / `INSERT <expr>`
 * expressions, allowing per-pair computation across both sides.
 *
 * Execution semantics:
 *
 * - The source RHS is evaluated once to produce a list of rows.
 * - For each source row, the target list is scanned for a matching
 *   row using the `ON` clause.
 * - If a match is found and a `WHEN MATCHED` branch is present, that
 *   branch is applied (rewriting the matched target row in place, or
 *   removing it).  Without a `WHEN MATCHED` branch, the source row is
 *   silently ignored.
 * - If no match is found and a `WHEN NOT MATCHED THEN INSERT` branch
 *   is present, the row is appended to the target list.  Without the
 *   branch, the source row is silently ignored.
 *
 * The target binding must already exist (use `LET` to create it).
 */
/**
 * Proxy AST node used by the parser to register the `target` /
 * `source` alias in the variable scope while parsing the ON
 * predicate and SET / INSERT expressions.  Resolves to the row
 * currently exposed by the owning Merge instance.
 */
export class MergeTargetAlias extends ASTNode {
    private _merge: Merge;
    constructor(merge: Merge) {
        super();
        this._merge = merge;
    }
    public value(): any {
        return this._merge.targetValue();
    }
}

export class MergeSourceAlias extends ASTNode {
    private _merge: Merge;
    constructor(merge: Merge) {
        super();
        this._merge = merge;
    }
    public value(): any {
        return this._merge.sourceValue();
    }
}

class Merge extends Operation {
    private _name: string;
    private _targetAlias: string | null;
    private _sourceAlias: string | null;
    private _onClause: MergeOnClause;
    private _matched: MergeMatchedAction | null;
    private _notMatched: MergeNotMatchedAction | null;
    private _sourceExpression: Expression | null;
    private _sourceSubQuery: ASTNode | null;

    /**
     * The target row currently in scope (during ON / SET / INSERT
     * evaluation).  `undefined` for `WHEN NOT MATCHED` rows.
     */
    private _currentTarget: any = undefined;
    /**
     * The source row currently in scope (during ON / SET / INSERT
     * evaluation).
     */
    private _currentSource: any = undefined;

    /** The last value written back to the binding (for `results`). */
    private _value: any[] = [];

    constructor(
        name: string,
        targetAlias: string | null,
        sourceAlias: string | null,
        sourceExpression: Expression | null,
        sourceSubQuery: ASTNode | null,
        onClause: MergeOnClause,
        matched: MergeMatchedAction | null,
        notMatched: MergeNotMatchedAction | null
    ) {
        super();
        this._name = name;
        this._targetAlias = targetAlias;
        this._sourceAlias = sourceAlias;
        this._sourceExpression = sourceExpression;
        this._sourceSubQuery = sourceSubQuery;
        this._onClause = onClause;
        this._matched = matched;
        this._notMatched = notMatched;
        if (sourceExpression !== null) {
            this.addChild(sourceExpression);
        }
        if (sourceSubQuery !== null) {
            this.addChild(sourceSubQuery);
        }
        if (onClause.type === "predicate") {
            this.addChild(onClause.predicate);
        }
        if (matched !== null && matched.type === "update") {
            for (const item of matched.setItems) {
                if (item.expression !== null) {
                    this.addChild(item.expression);
                }
            }
        }
        if (notMatched !== null && notMatched.expression !== null) {
            this.addChild(notMatched.expression);
        }
    }

    /**
     * Replaces the placeholder ON clause and matched / not-matched
     * actions with their parsed counterparts (used by the parser,
     * which must register the operation in its variable scope before
     * parsing those expressions so that the target / source aliases
     * resolve to this node).
     */
    public setClauses(
        onClause: MergeOnClause,
        matched: MergeMatchedAction | null,
        notMatched: MergeNotMatchedAction | null
    ): void {
        this._onClause = onClause;
        this._matched = matched;
        this._notMatched = notMatched;
        if (onClause.type === "predicate") {
            this.addChild(onClause.predicate);
        }
        if (matched !== null && matched.type === "update") {
            for (const item of matched.setItems) {
                if (item.expression !== null) {
                    this.addChild(item.expression);
                }
            }
        }
        if (notMatched !== null && notMatched.expression !== null) {
            this.addChild(notMatched.expression);
        }
    }

    public get name(): string {
        return this._name;
    }

    public get targetAlias(): string | null {
        return this._targetAlias;
    }

    public get sourceAlias(): string | null {
        return this._sourceAlias;
    }

    public get sourceExpression(): Expression | null {
        return this._sourceExpression;
    }

    public get sourceSubQuery(): ASTNode | null {
        return this._sourceSubQuery;
    }

    public get onClause(): MergeOnClause {
        return this._onClause;
    }

    public get matched(): MergeMatchedAction | null {
        return this._matched;
    }

    public get notMatched(): MergeNotMatchedAction | null {
        return this._notMatched;
    }

    /**
     * Returns the row currently in scope for the given alias.  This is
     * what {@link Reference}.value() resolves to when the target /
     * source alias is referenced inside the ON predicate or in a
     * SET / INSERT expression.  The parser stamps an aliasContext on
     * the proxy AST nodes it registers in the variable scope so we
     * can disambiguate without name comparison.
     */
    public targetValue(): any {
        return this._currentTarget;
    }

    public sourceValue(): any {
        return this._currentSource;
    }

    public async run(): Promise<void> {
        const bindings = Bindings.getInstance();
        if (!bindings.has(this._name)) {
            throw new Error(`Binding '${this._name}' is not defined; use LET to create it`);
        }
        const existing = bindings.get(this._name);
        if (!Array.isArray(existing)) {
            throw new Error(`MERGE INTO ${this._name} requires '${this._name}' to be a list`);
        }
        // Evaluate source once.
        let source: any;
        if (this._sourceSubQuery !== null) {
            const first = this._sourceSubQuery.firstChild() as Operation;
            const last = this._sourceSubQuery.lastChild() as Operation;
            await first.initialize();
            await first.run();
            await first.finish();
            source = last.results;
        } else if (this._sourceExpression !== null) {
            source = this._sourceExpression.value();
        } else {
            source = null;
        }
        if (!Array.isArray(source)) {
            throw new Error(
                `MERGE INTO ${this._name} USING … requires the source to evaluate to a list of rows`
            );
        }

        // Working copy of the target list; mutations are applied in
        // place and a final assignment writes it back to the binding.
        const result: any[] = existing.slice();
        // Tombstone marker for delete actions; we sweep at the end so
        // that intra-merge index lookups remain stable.
        const tombstones: Set<number> = new Set();

        for (const sourceRow of source) {
            this._currentSource = sourceRow;
            const matchedIndex = this.findMatch(result, tombstones, sourceRow);
            if (matchedIndex !== -1) {
                this._currentTarget = result[matchedIndex];
                if (this._matched !== null) {
                    if (this._matched.type === "delete") {
                        tombstones.add(matchedIndex);
                    } else {
                        result[matchedIndex] = this.applySet(
                            result[matchedIndex],
                            this._matched.setItems
                        );
                    }
                }
                this._currentTarget = undefined;
            } else {
                if (this._notMatched !== null) {
                    const inserted =
                        this._notMatched.expression !== null
                            ? this._notMatched.expression.value()
                            : sourceRow;
                    result.push(inserted);
                }
            }
        }
        this._currentSource = undefined;

        const swept: any[] = [];
        for (let i = 0; i < result.length; i++) {
            if (!tombstones.has(i)) {
                swept.push(result[i]);
            }
        }
        bindings.set(this._name, swept);
        this._value = swept;
        await this.next?.run();
    }

    /**
     * Finds the first index in `target` (skipping tombstones) for
     * which the ON clause holds against the current source row.  The
     * source row is already exposed via {@link sourceValue}.
     */
    private findMatch(target: any[], tombstones: Set<number>, sourceRow: any): number {
        if (this._onClause.type === "keys") {
            const keys = this._onClause.keys;
            // Source row must be an object with all keys defined; if not,
            // it cannot match anything and falls through to NOT MATCHED.
            if (sourceRow === null || typeof sourceRow !== "object") {
                return -1;
            }
            for (const k of keys) {
                if (!(k in sourceRow)) {
                    return -1;
                }
            }
            for (let i = 0; i < target.length; i++) {
                if (tombstones.has(i)) {
                    continue;
                }
                const t = target[i];
                if (t === null || typeof t !== "object") {
                    continue;
                }
                let matches = true;
                for (const k of keys) {
                    if (!(k in t) || t[k] !== sourceRow[k]) {
                        matches = false;
                        break;
                    }
                }
                if (matches) {
                    return i;
                }
            }
            return -1;
        }
        // Predicate ON — evaluate per (target, source) pair.
        for (let i = 0; i < target.length; i++) {
            if (tombstones.has(i)) {
                continue;
            }
            this._currentTarget = target[i];
            const ok = this._onClause.predicate.value();
            if (ok) {
                this._currentTarget = undefined;
                return i;
            }
        }
        this._currentTarget = undefined;
        return -1;
    }

    /**
     * Builds the next value for a matched target row by applying the
     * `SET <set-list>` items.  Items without an explicit expression
     * default to `source.<field>`.
     */
    private applySet(targetRow: any, setItems: MergeSetItem[]): any {
        const base = targetRow !== null && typeof targetRow === "object" ? { ...targetRow } : {};
        for (const item of setItems) {
            if (item.expression !== null) {
                base[item.field] = item.expression.value();
            } else {
                // Default: copy source.<field> if present, otherwise
                // leave the existing field untouched.  This makes
                // `SET .name, .email` behave like the old partial-merge
                // form (only listed fields overwrite; missing ones
                // preserve the target's value).
                const src = this._currentSource;
                if (src !== null && typeof src === "object" && item.field in src) {
                    base[item.field] = src[item.field];
                }
            }
        }
        return base;
    }

    public get results(): Record<string, any>[] {
        return this._value as Record<string, any>[];
    }
}

export default Merge;

File: flowquery-py/src/parsing/operations/merge.py
"""MERGE INTO … USING … — per-row upsert operation."""

from typing import Any, Dict, List, Optional, Set, Union, cast

from ...graph.bindings import Bindings
from ..ast_node import ASTNode
from ..expressions.expression import Expression
from .operation import Operation


class MergeSetItem:
    """One item in the ``WHEN MATCHED THEN UPDATE SET`` list.

    ``field`` is the name of the field on the matched target row to
    overwrite; ``expression`` is the value to assign.  When
    ``expression`` is ``None``, the default is ``source.<field>`` —
    i.e. the value of the same-named field on the current incoming
    row.  This makes the partial-merge case (``SET .name, .email``)
    ergonomic without having to write
    ``SET .name = source.name, .email = source.email``.
    """

    def __init__(self, field: str, expression: Optional[Expression]):
        self.field = field
        self.expression = expression


class MergeMatchedUpdate:
    """``WHEN MATCHED THEN UPDATE SET <items>`` action."""

    type: str = "update"

    def __init__(self, set_items: List[MergeSetItem]):
        self.set_items = set_items


class MergeMatchedDelete:
    """``WHEN MATCHED THEN DELETE`` action."""

    type: str = "delete"


MergeMatchedAction = Union[MergeMatchedUpdate, MergeMatchedDelete]


class MergeNotMatchedInsert:
    """``WHEN NOT MATCHED THEN INSERT [<expr>]`` action.  ``expression``
    is ``None`` for the bare-insert default (the source row itself).
    """

    type: str = "insert"

    def __init__(self, expression: Optional[Expression]):
        self.expression = expression


class MergeOnKeys:
    """``ON <key>`` or ``ON (<k1>, <k2>, …)`` — positional key match."""

    type: str = "keys"

    def __init__(self, keys: List[str]):
        self.keys = keys


class MergeOnPredicate:
    """``ON <predicate>`` — arbitrary per-pair predicate."""

    type: str = "predicate"

    def __init__(self, predicate: Expression):
        self.predicate = predicate


MergeOnClause = Union[MergeOnKeys, MergeOnPredicate]


class MergeTargetAlias(ASTNode):
    """Proxy AST node used by the parser to register the target alias
    in the variable scope while parsing the ON predicate and SET /
    INSERT expressions.  Resolves to the row currently exposed by the
    owning Merge instance.
    """

    def __init__(self, merge: "Merge"):
        super().__init__()
        self._merge = merge

    def value(self) -> Any:
        return self._merge.target_value()


class MergeSourceAlias(ASTNode):
    """Proxy AST node for the source alias (see :class:`MergeTargetAlias`)."""

    def __init__(self, merge: "Merge"):
        super().__init__()
        self._merge = merge

    def value(self) -> Any:
        return self._merge.source_value()


class Merge(Operation):
    """``MERGE INTO <name> [AS <target>] USING <source-rhs> [AS <source>]
    ON ( <key> | (<k1>, <k2>, …) | <predicate> )
    [ WHEN MATCHED THEN UPDATE SET <set-list> ]
    [ WHEN MATCHED THEN DELETE ]
    [ WHEN NOT MATCHED THEN INSERT [<expr>] ]``

    SQL-MERGE-style upsert into an existing list-binding.  The source
    may be any expression, sub-query, or bare binding name; both the
    ``target`` and ``source`` aliases are in scope inside the ``ON``
    predicate and any ``SET .field = <expr>`` / ``INSERT <expr>``
    expressions, allowing per-pair computation across both sides.
    """

    def __init__(
        self,
        name: str,
        target_alias: Optional[str],
        source_alias: Optional[str],
        source_expression: Optional[Expression],
        source_sub_query: Optional[ASTNode],
        on_clause: MergeOnClause,
        matched: Optional[MergeMatchedAction],
        not_matched: Optional[MergeNotMatchedInsert],
    ):
        super().__init__()
        self._name = name
        self._target_alias = target_alias
        self._source_alias = source_alias
        self._source_expression = source_expression
        self._source_sub_query = source_sub_query
        self._on_clause = on_clause
        self._matched = matched
        self._not_matched = not_matched
        self._current_target: Any = None
        self._current_source: Any = None
        self._value: List[Any] = []
        if source_expression is not None:
            self.add_child(source_expression)
        if source_sub_query is not None:
            self.add_child(source_sub_query)
        # `on_clause` / `matched` / `not_matched` children are wired up
        # by `set_clauses`, which the parser calls after parsing those
        # expressions in the alias-augmented scope.

    def set_clauses(
        self,
        on_clause: MergeOnClause,
        matched: Optional[MergeMatchedAction],
        not_matched: Optional[MergeNotMatchedInsert],
    ) -> None:
        """Replace placeholder ON / matched / not-matched clauses with
        their parsed counterparts (used by the parser, which must
        register the Merge in its variable scope before parsing the
        expressions inside these clauses so that target / source
        aliases resolve to this node)."""
        self._on_clause = on_clause
        self._matched = matched
        self._not_matched = not_matched
        if isinstance(on_clause, MergeOnPredicate):
            self.add_child(on_clause.predicate)
        if isinstance(matched, MergeMatchedUpdate):
            for item in matched.set_items:
                if item.expression is not None:
                    self.add_child(item.expression)
        if not_matched is not None and not_matched.expression is not None:
            self.add_child(not_matched.expression)

    @property
    def name(self) -> str:
        return self._name

    @property
    def target_alias(self) -> Optional[str]:
        return self._target_alias

    @property
    def source_alias(self) -> Optional[str]:
        return self._source_alias

    @property
    def source_expression(self) -> Optional[Expression]:
        return self._source_expression

    @property
    def source_sub_query(self) -> Optional[ASTNode]:
        return self._source_sub_query

    @property
    def on_clause(self) -> MergeOnClause:
        return self._on_clause

    @property
    def matched(self) -> Optional[MergeMatchedAction]:
        return self._matched

    @property
    def not_matched(self) -> Optional[MergeNotMatchedInsert]:
        return self._not_matched

    def target_value(self) -> Any:
        return self._current_target

    def source_value(self) -> Any:
        return self._current_source

    async def run(self) -> None:
        bindings = Bindings.get_instance()
        if not bindings.has(self._name):
            raise RuntimeError(
                f"Binding '{self._name}' is not defined; use LET to create it"
            )
        if bindings.is_refreshable(self._name):
            raise RuntimeError(
                f"Binding '{self._name}' is refreshable; use REFRESH BINDING "
                f"{self._name} to re-evaluate or DROP BINDING {self._name} first"
            )
        existing = bindings.get(self._name)
        if not isinstance(existing, list):
            raise RuntimeError(
                f"MERGE INTO {self._name} requires '{self._name}' to be a list"
            )
        # Evaluate source once.
        source: Any
        if self._source_sub_query is not None:
            first = cast(Operation, self._source_sub_query.first_child())
            last = cast(Operation, self._source_sub_query.last_child())
            await first.initialize()
            await first.run()
            await first.finish()
            source = last.results
        elif self._source_expression is not None:
            source = self._source_expression.value()
        else:
            source = None
        if not isinstance(source, list):
            raise RuntimeError(
                f"MERGE INTO {self._name} USING … requires the source to "
                "evaluate to a list of rows"
            )

        result: List[Any] = list(existing)
        tombstones: Set[int] = set()

        for source_row in source:
            self._current_source = source_row
            matched_index = self._find_match(result, tombstones, source_row)
            if matched_index != -1:
                self._current_target = result[matched_index]
                if self._matched is not None:
                    if isinstance(self._matched, MergeMatchedDelete):
                        tombstones.add(matched_index)
                    else:
                        result[matched_index] = self._apply_set(
                            result[matched_index], self._matched.set_items
                        )
                self._current_target = None
            else:
                if self._not_matched is not None:
                    inserted = (
                        self._not_matched.expression.value()
                        if self._not_matched.expression is not None
                        else source_row
                    )
                    result.append(inserted)
        self._current_source = None

        swept: List[Any] = [
            r for i, r in enumerate(result) if i not in tombstones
        ]
        bindings.set(self._name, swept)
        self._value = swept
        if self.next:
            await self.next.run()

    def _find_match(
        self, target: List[Any], tombstones: Set[int], source_row: Any
    ) -> int:
        if isinstance(self._on_clause, MergeOnKeys):
            keys = self._on_clause.keys
            if not isinstance(source_row, dict):
                return -1
            for k in keys:
                if k not in source_row:
                    return -1
            for i, t in enumerate(target):
                if i in tombstones:
                    continue
                if not isinstance(t, dict):
                    continue
                matches = True
                for k in keys:
                    if k not in t or t[k] != source_row[k]:
                        matches = False
                        break
                if matches:
                    return i
            return -1
        predicate = self._on_clause.predicate
        for i, t in enumerate(target):
            if i in tombstones:
                continue
            self._current_target = t
            ok = predicate.value()
            if ok:
                self._current_target = None
                return i
        self._current_target = None
        return -1

    def _apply_set(
        self, target_row: Any, set_items: List[MergeSetItem]
    ) -> Dict[str, Any]:
        base: Dict[str, Any] = (
            dict(target_row) if isinstance(target_row, dict) else {}
        )
        for item in set_items:
            if item.expression is not None:
                base[item.field] = item.expression.value()
            else:
                src = self._current_source
                if isinstance(src, dict) and item.field in src:
                    base[item.field] = src[item.field]
        return base

    @property
    def results(self) -> List[Any]:
        return self._value
